be/src/io/async_flush_output_stream.cpp

```
@@ -0,0 +1,161 @@
+                delete chunk;
+            });
+            auto status = _file->append(Slice(buffer->data(), buffer->size()));
+            {
+                std::scoped_lock lock(_mutex);
+                _io_status.update(status);
+                if (_task_queue.empty()) {
+                    _has_in_flight_io = false;
+                    return;
+                }
+                auto task = _task_queue.front();
+                _task_queue.pop();
+                bool ok = _io_executor->offer(task);
+                DCHECK(ok);
+            }
+        };
+        to_enqueue_tasks.emplace_back(task);
+    }
+
+    auto close_task = [&]() {
+        SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER(_runtime_state->instance_mem_tracker());
+        CurrentThread::current().set_query_id(_runtime_state->query_id());
+        CurrentThread::current().set_fragment_instance_id(_runtime_state->fragment_instance_id());
+        auto status = _file->close();
+        {
+            std::scoped_lock lock(_mutex);
+            _io_status.update(status);
+            DCHECK(_task_queue.empty()); // close task is the last task
+            _has_in_flight_io = false;
+            _promise.set_value(_io_status); // notify
+        }
+    };
+    to_enqueue_tasks.emplace_back(close_task);
+    enqueue_tasks_and_maybe_submit_task(std::move(to_enqueue_tasks));
+    return Status::OK();
+}
+
+void AsyncFlushOutputStream::enqueue_tasks_and_maybe_submit_task(std::vector<Task> tasks) {
+    if (_io_executor == nullptr) {
+        for (auto& task : tasks) {
+            task();
+        }
+        return;
+    }
+
+    std::scoped_lock lock(_mutex);
+    std::for_each(tasks.begin(), tasks.end(), [&](auto& task) { _task_queue.push(task); });
+
+    if (_has_in_flight_io) {
+        return;
+    }
+    auto task = _task_queue.front();
+    _task_queue.pop();
+    _has_in_flight_io = true;
+    bool ok = _io_executor->offer(task);
+    DCHECK(ok);
```

ditto, for sink, ignore exception is dangerous